TargetComposition/SymBuilder/ApiProvider.cpp

```diff
@@ -2172,6 +2265,307 @@ void LiveRangeObject::Delete(_In_ const Object& /*liveRangeObject*/,
+    }
+}
+
+std::wstring PublicObject::ToString(_In_ const Object& /*publicObject*/,
+                                    _In_ ComPtr<PublicSymbol>& spPublicSymbol,
+                                    _In_ const Metadata& /*metadata*/)
+{
+    std::wstring const& publicName = spPublicSymbol->InternalGetQualifiedName();
+
+    std::wstring displayString = L"Public Symbol: ";
+    displayString += (publicName.empty() ? L"<Unknown>" : publicName.c_str());
+
+    wchar_t buf[128];
+    swprintf_s(buf, ARRAYSIZE(buf), L" (offset = 0x%I64x)", spPublicSymbol->InternalGetOffset());
+
+    displayString += buf;
+
+    return displayString;
+}
+
+ULONG64 PublicObject::GetOffset(_In_ const Object& /*publicObject*/, 
+                                _In_ ComPtr<PublicSymbol>& spPublicSymbol)
+{
+    return spPublicSymbol->InternalGetOffset();
+}
+
+void PublicObject::Delete(_In_ const Object& /*publicObject*/, 
+                          _In_ ComPtr<PublicSymbol>& spPublicSymbol)
+{
+    CheckHr(spPublicSymbol->Delete());
+}
+
+Object PublicObject::PromoteToFunction(_In_ const Object& publicObject, 
+                                       _In_ ComPtr<PublicSymbol>& spPublicSymbol,
+                                       _In_ std::optional<ULONG64> codeSize,
+                                       _In_ std::optional<Object> returnType,
+                                       _In_ size_t argCount,                 // [parameter]...
+                                       _In_reads_(argCount) Object *pArgs)
+{
+    SymbolSet *pSymbolSet = spPublicSymbol->InternalGetSymbolSet();
+    if (!returnType.has_value())
+    {
+        returnType = L"void";
+    }
+
+    BaseTypeSymbol *pReturnType = UnboxType(pSymbolSet, returnType);
+
+    //
+    // If the code size was specified as zero (or was left out of the argument list), we will go down
+    // to the disassembler and ask it to do some work for us.  Note that the data model disassembler is provided
+    // by another key extension to the debugger (the API extension).  If that extension isn't available,
+    // accessing these properties will throw (and the promotion will fail).
+    //
+    if (codeSize.value_or(0) == 0)
+    {
+        ULONG64 codeEntryVa = spPublicSymbol->InternalGetOffset();
+        ISvcModule *pModule = pSymbolSet->GetModule();
+
+        ULONG64 moduleBase;
+        CheckHr(pModule->GetBaseAddress(&moduleBase));
+        codeEntryVa += moduleBase;
+
+        //
+        // Many of the debugger's "API" type of extensions are somewhere under Debugger.Utility.  The data model
+        // disassembler places itself under "Debugger.Utility.Code".
+        //
+        Object codeNamespace = Object::RootNamespace().KeyValue(L"Debugger")
+                                                      .KeyValue(L"Utility")
+                                                      .KeyValue(L"Code");
+
+        Object disassembler = codeNamespace.CallMethod(L"CreateDisassembler");

```

Comment:
What happens if DbgModelApiXtn isn't available? I assume it's a C++ exception and this function just returns failure, right?

Reply:
Nevermind, read your comment closer.

Reply:
Yes -- you'll get a C++ exception thrown from KeyValue.  Bound properties/methods with the library are fine to throw exceptions...  at the data model level, you'll get a failure from IModelMethod::Call.